pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java

```diff
@@ -31,152 +31,90 @@
-    _stopPeriodicTask = true;
-
-    LOGGER.info("Waiting for periodic task {} to finish, maxWaitTimeMillis = {}", getTaskName(),
-        MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS);
-    long millisToWait = MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS;
-    while (_periodicTaskInProgress && millisToWait > 0) {
-      try {
-        long thisWait = 1000;
-        if (millisToWait < thisWait) {
-          thisWait = millisToWait;
-        }
-        Thread.sleep(thisWait);
-        millisToWait -= thisWait;
-      } catch (InterruptedException e) {
-        LOGGER.info("Interrupted: Remaining wait time {} (out of {}) for task {}", millisToWait,
-            MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS, getTaskName());
+    LOGGER.info("Processing {} tables in task: {}", numTables, _taskName);
+    C context = preprocess();
+    int numTablesProcessed = 0;
+    for (String tableNameWithType : tableNamesWithType) {
+      if (!isStarted()) {
+        LOGGER.info("Task: {} is stopped, early terminate the task", _taskName);
         break;
       }
-    }
-    LOGGER.info("Wait completed for task {}. Waited for {} ms. _periodicTaskInProgress = {}", getTaskName(),
-        MAX_CONTROLLER_PERIODIC_TASK_STOP_TIME_MILLIS - millisToWait, _periodicTaskInProgress);
-
-    stopTask();
-  }
-
-  /**
-   * Processes the task on the given tables.
-   *
-   * @param tableNamesWithType List of table names
-   */
-  protected void process(List<String> tableNamesWithType) {
-    if (!shouldStopPeriodicTask()) {
-
-      int numTablesProcessed = 0;
-      preprocess();
-
-      for (String tableNameWithType : tableNamesWithType) {
-        if (shouldStopPeriodicTask()) {
-          LOGGER.info("Skip processing table {} and all the remaining tables for task {}.", tableNameWithType,
-              getTaskName());
-          break;
-        }
-        try {
-          processTable(tableNameWithType);
-          numTablesProcessed++;
-        } catch (Exception e) {
-          exceptionHandler(tableNameWithType, e);
-        }
+      try {
+        processTable(tableNameWithType, context);
+      } catch (Exception e) {
+        LOGGER.error("Caught exception while processing table: {} in task: {}", tableNamesWithType, _taskName, e);
       }
-
-      postprocess();
-      _metricsRegistry
-          .setValueOfGlobalGauge(ControllerGauge.PERIODIC_TASK_NUM_TABLES_PROCESSED, getTaskName(), numTablesProcessed);
-    } else {
-      LOGGER.info("Skip processing all tables for task {}", getTaskName());
+      numTablesProcessed++;
     }
+    postprocess(context);
+    _controllerMetrics
+        .setValueOfGlobalGauge(ControllerGauge.PERIODIC_TASK_NUM_TABLES_PROCESSED, _taskName, numTablesProcessed);
+    LOGGER.info("Finish processing {}/{} tables in task: {}", numTablesProcessed, numTables, _taskName);
   }
 
   /**
-   * This method runs before processing all tables
+   * Can be override to provide context before processing the tables.
```

Comment:
*can be overridden

Reply:
Fixed

Reply:
I think you missed it again :) Here and also in 2 other places in this file